File: 0127-word-ladder/0127-word-ladder.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
         
         queue<pair<string,int>>q;
         unordered_set<string>st(wordList.begin(),wordList.end());
         q.push({beginWord,1});
         st.erase(beginWord);

         while(!q.empty()){
            string word = q.front().first;
            int len = q.front().second;
            q.pop();
            if(word == endWord){
              return len;
            }

            for(int i=0;i<word.size();i++){
               char original = word[i];

               for(char c = 'a';c<='z';c++){
                 word[i] = c;
                 if(st.find(word) != st.end()){
                    st.erase(word);
                    q.push({word,len+1});
                 }
               }

               word[i] = original;

            }
         }
         return 0;
    }
};

File: 0127-word-ladder/0127-word-ladder.java
class Solution {

    class Pair{
        String word;
        int freq;

        Pair(String word, int freq){
           this.word = word;
           this.freq = freq;
        }
    }
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Queue<Pair> q = new LinkedList<>();
        Set<String> st = new HashSet<>();
        for(int i=0;i<wordList.size();i++){
           st.add(wordList.get(i));
        }
        q.add(new Pair(beginWord,1));
        st.remove(beginWord);
        while(q.size() > 0){
            Pair p = q.poll();
            String word = p.word;
            int freq = p.freq;
           if(word.equals(endWord) == true) return freq;
            char[] word_arr = word.toCharArray();
            for(int i=0;i<word.length();i++){
               for(char c = 'a';c<= 'z';c++){
                 word_arr[i] = c;
                 String str = new String(word_arr);
                 if(st.contains(str) == true){
                    st.remove(str);
                    q.add(new Pair(str, freq+1));
                    
                 }
               }
               word_arr = word.toCharArray();
            }
        }

        return 0;
    }
}